src/wallet/sqlite.cpp

```diff
@@ -15,16 +15,42 @@
 
 static const char* DATABASE_FILENAME = "wallet.dat";
 
+std::atomic<int> g_dbs_open{0};
+
+static void ErrorLogCallback(void* arg, int code, const char* msg)
+{
+    assert(arg == nullptr); // That's what we tell it to do during the setup
```

Comment:
> It feels strange to me that the log callback makes everything blow up in a special case. I think we use asserts for developer errors but I don't think that's the case here.

This would only be hit in a developer error or a sqlite bug.

> Also doesn't this mean that the `throw std::runtime_error` on L36 will never be used?

I don't think so?

Reply:
Hm, the docs aren't very clear on this (or I can't find the right place) so I can not say yet if there are cases where `ret == SQLITE_OK` but the log callback is still being called. That would be the only case when this seems valuable because otherwise the `runtime_error` would be hit in case there is a problem and the error message would be clearer I think? Yes, this could be a bug in SQLite but in general, we are trusting that the ret/res codes are not lying so I don't know why we would do it differently here.

On the flip side, the `runtime_error` will currently only be hit if `ret != SQLITE_OK` but the `ErrorLogCallback` is not being called. Again, I couldn't find details/guarantees on this in the docs but at least they say the callback will be used ['whenever anomalies occur'](https://www.sqlite.org/errlog.html), so my expectation is that it would always be called as whenever `ret != SQLITE_OK` is true. So I think the `runtime_error` may never be hit unless there is a bug in SQLite.

Overall it just feels strange to do anything but logging in a function that is called `*LogCallback`.

Reply:
I think if `ret != SQLITE_OK` both `ErrorLogCallback` and `runtime_error` get called. The `ErrorLogCallback` doesn't change `ret` to `SQLITE_OK` and it doesn't throw it's own exception. The assertion should never fail so that won't cause a program exit either.

Reply:
Added this comment.